### PR TITLE
Make `Raster.to_pointcloud()` modular and add `Raster.get_mask()` for memory-efficient operations requiring mask

### LINKED ISSUE #499: Allow user to pass `z_name` and `random_state` for `Raster.to_points()`
Right now forced to "b1" without any argument.
Could rename the method into `to_pointcloud` for clarity?

### LINKED ISSUE #502: Operations requiring only `Raster.data.mask` load the entire data array
We can use `rasterio.read_masks()` to only get the mask and perform operations requiring only the data mask (and not the data array) in a much more memory-efficient way.
This includes for example:

- `subsample()` and any functionality that uses it like `to_points()`,
- `crop()` with option to crop only to nodata (not yet implemented, see #400).

Starting a `_load_only_mask` function in #501...

### LINKED ISSUE #504: `subsample_array` still runs random subsampling with `subsample=1`
Effectively "shuffles" the dataset, I think it is better to skip the subsampling entirely... Especially for testing purposes.

Ongoing in #501

### LINKED ISSUE #505: Deal consistently with `Raster.data.mask=False`
The mask of a masked array can be just `False` and not an array of `False` when there is no nodata. This can lead to some problems in functions.
The function `geoutils.raster.array.get_mask` circumvents that issue by using `np.isfinite(array)` on the array if the mask is not an array, but that requires the array (which we wouldn't have necessarily loaded, see #502) and mostly more CPU usage.

We should have a consistent way to deal with this, maybe something similar to what is done in `raster_equal` here: https://github.com/GlacioHack/geoutils/blob/d0cd168507a80f7c240fc075c8d856d7dbe6a2a4/geoutils/raster/raster.py#L988
Or a "good practice" from the NumPy documentation?

### PR DESCRIPTION
**Pull request by rhugonnet**
## Summary
This PR improves `Raster.to_pointcloud()` in terms of argument consistency, modularity, and makes its subsampling functionality only returns valid values (while before it returned any random sample) to also be consistent with other subsampling methods.
Additionally, this PR adds a non-public `Raster._load_only_mask()` function called by a new public function `Raster.get_mask()`, which allows to get the nodata mask without loading the array into memory. This is used in `.to_pointcloud()` to perform sampling of only valid values without loading the array in memory :smile:.

The PR also adds a lot of tests!

## Detailed changes

This PR:
- Renames `Raster.to_points()` into `Raster.to_pointcloud()` for consistent naming in preparation for #492,
- Define a main `data_column_name` argument for `Raster.to_pointcloud()` to associate the point cloud to a certain `data_band` (instead of having a hard-coded "b1", "b2", etc), and proposes to store all columns using the new argument `store_auxiliary_bands` (naming to be made consistent with `PointCloud` class but should be fairly close),
- Adds a `random_state` argument for `subsample`,
- Changes the subsampling done to rely on `subsample_array()`, after calling a new `_load_only_mask` function to get the valid mask from disk without loading the raster array,
- Modifies `_load_rio` to allow to return only the mask by calling `rasterio.read_masks()`,
- Adds a `Raster.get_mask()` function that returns the raster mask consistently as a boolean array (to circumvent the issue of having a single `False` value, see #505) and without loading the raster array.

Resolves #499
Resolves #504
Resolves #505
Resolves #502 

TO-DO:

- [x] Add tests for `subsample=1` in `subsample_array()`,
- [x] Add tests for `_load_only_mask`,
- [x] Define a function `get_mask` that works without the array? See #505.

### CODE REVIEW

File: examples/handling/interface/topoints.py [hunk 1]
```diff
@@ -21,10 +21,10 @@
...
 
-pts_rast = rast.to_points()
+pts_rast = rast.to_pointcloud()
 pts_rast
 
 # %%
 # We plot the point vector.
 
-pts_rast.plot(column="b1", cmap="terrain", legend=True)
+pts_rast.ds.plot(column="b1", cmap="terrain", legend=True)
```
Comment (adehecq) [inline examples/handling/interface/topoints.py:30]: Why do we need to add `ds`? Isn't `pts_rast` a vector instance? In which case, the method `plot` should work directly...

File: geoutils/raster/raster.py [hunk 2]
```diff
@@ -3363,96 +3443,186 @@ def split_bands(self: RasterType, copy: bool = False, bands: list[int] | int | N
...
 
-        If the raster is not loaded, sampling will be done from disk without loading the entire Raster.
+        If the raster is not loaded, sampling will be done from disk using rasterio.sample after loading only the masks
+        of the dataset.
 
         Formats:
             * `as_array` == False: A vector with dataframe columns ["b1", "b2", ..., "geometry"],
             * `as_array` == True: A numpy ndarray of shape (N, 2 + count) with the columns [x, y, b1, b2..].
 
+        :param data_column_name: Name of point cloud data column to use, defaults to "b1".
```
Comment (adehecq) [inline geoutils/raster/raster.py:3532]: I would rephrase: "Name to use for point cloud data column" or it sounds like we need to specify the name from some inputs.

File: geoutils/raster/raster.py [hunk 3]
```diff
@@ -3363,96 +3443,186 @@ def split_bands(self: RasterType, copy: bool = False, bands: list[int] | int | N
...
+        random_state: np.random.RandomState | int | None = None,
+        force_pixel_offset: Literal["center", "ul", "ur", "ll", "lr"] = "ul",
+    ) -> NDArrayNum | Vector:
+        ...
+
+    def to_pointcloud(
+        self,
+        data_column_name: str = "b1",
+        data_band: int = 1,
+        store_auxiliary_bands: bool = False,
```
Comment (adehecq) [inline geoutils/raster/raster.py:3500]: One question that comes up. For those bands, the name will be "bX" whatever the name chosen for `data_column_name`? Or would it make sense to either use a prefix, or a list of names for multibands?

File: geoutils/raster/raster.py [hunk 4]
```diff
@@ -3363,96 +3443,186 @@ def split_bands(self: RasterType, copy: bool = False, bands: list[int] | int | N
...
         :param as_array: Return an array instead of a vector.
-        :param pixel_offset: The point at which to associate the pixel coordinate with ('corner' == upper left).
+        :param random_state: Random state or seed number.
+        :param force_pixel_offset: Force offset to derive point coordinate with. Raster coordinates normally only
+            associate to upper-left corner "ul" ("Area" definition) or center ("Point" definition).
 
         :raises ValueError: If the sample count or fraction is poorly formatted.
 
-        :returns: A GeoDataFrame, or ndarray of the shape (N, 2 + count) where N is the sample count.
+        :returns: A point cloud, or ndarray of the shape (N, 2 + count) where N is the sample count.
```
Comment (adehecq) [inline geoutils/raster/raster.py:3545]: It is not clear what a "point cloud" is.

File: geoutils/raster/raster.py [hunk 5]
```diff
@@ -3363,96 +3443,186 @@ def split_bands(self: RasterType, copy: bool = False, bands: list[int] | int | N
...
-            sample = int(data_size * subsample)
-        else:
-            sample = int(subsample)
-        if sample > data_size:
-            raise ValueError(f"sample cannot exceed the size of the dataset ({sample} vs {data_size})")
+        # Input checks
+        if not isinstance(data_column_name, str):
+            raise ValueError("Data column name must be a string.")
+        if not isinstance(data_band, int) and (1 < data_band or self.count < data_band):
+            raise ValueError(f"Data band number must be between 1 and the total number of bands of {self.count}.")
```
Comment (adehecq) [inline geoutils/raster/raster.py:3552]: Rephrase "of {self.count}" into "({self.count})".

File: geoutils/raster/raster.py [hunk 6]
```diff
@@ -3363,96 +3443,186 @@ def split_bands(self: RasterType, copy: bool = False, bands: list[int] | int | N
...
+        ma_valid = np.ma.masked_array(data=np.ones(np.shape(valid_mask), dtype=bool), mask=~valid_mask)
+        # Take a subsample within the valid values
+        indices = subsample_array(array=ma_valid, subsample=subsample, random_state=random_state, return_indices=True)
 
-        # Extract the coordinates of the pixels and filter by the chosen pixels.
-        x_coords, y_coords = (np.array(a) for a in self.ij2xy(rows, cols, offset=pixel_offset))
+        # Extract the coordinates at subsampled pixels with valid data
+        # To extract data, we always use "upper left" which rasterio interprets as the exact coordinates of the raster
+        # Further below we redefine output coordinates based on point interpretation
+        x_coords, y_coords = (np.array(a) for a in self.ij2xy(indices[0], indices[1], offset="ul"))
```
Comment (adehecq) [inline geoutils/raster/raster.py:3581]: It seems like `x_coords` and `y_coords` are only used in the second block of the following if statement, so could be moved there no?

File: tests/test_raster.py [hunk 7]
```diff
@@ -2966,51 +2984,122 @@ def test_proximity_parameters(self) -> None:
...
+        assert isinstance(points_arr, np.ndarray)
 
-        assert img0.to_points(0.2, as_array=True).shape == (5, 3)
+        # Check that both outputs (array or vector) are fully consistent, order matters here
+        assert np.array_equal(points.ds.geometry.x.values, points_arr[:, 0])
+        assert np.array_equal(points.ds.geometry.y.values, points_arr[:, 1])
+        assert np.array_equal(points.ds["b1"].values, points_arr[:, 2])
+
+        # Validate that 25 points were sampled (equating to img1.height * img1.width) with x, y, and band0 values.
+        assert isinstance(points_arr, np.ndarray)
```
Comment (adehecq) [inline tests/test_raster.py:3010]: this was already tested at L 3002.

File: tests/test_raster.py [hunk 8]
```diff
@@ -2966,51 +2984,122 @@ def test_proximity_parameters(self) -> None:
...
+        assert points_arr.shape == (25, 3)
+        assert points.ds.shape == (25, 2)  # One less column here due to geometry storing X and Y
+        # Check that X, Y and Z arrays are equal to raster array input independently of value order
+        x_coords, y_coords = img0.ij2xy(i=np.arange(0, 5), j=np.arange(0, 5))
```
Comment (adehecq) [inline tests/test_raster.py:3014]: can't you just use `Raster.coords`? https://geoutils.readthedocs.io/en/latest/gen_modules/geoutils.Raster.coords.html

**Review by adehecq — APPROVED** [review]
Apart some minor comments, looking good to me. Thanks !!

File: examples/handling/interface/topoints.py (hunk 1 above)
Comment (rhugonnet) [reply examples/handling/interface/topoints.py:30]: Indeed, changed back to `plot`!

File: geoutils/raster/raster.py (hunk 2 above)
Comment (rhugonnet) [reply geoutils/raster/raster.py:3532]: Done

File: geoutils/raster/raster.py (hunk 3 above)
Comment (rhugonnet) [reply geoutils/raster/raster.py:3500]: You're right! Just took some more time and extended the functionality to support any `auxiliary_data_band` as a list and `auxiliary_column_names` as a list too. Now the user can control everything! :smile:

File: geoutils/raster/raster.py (hunk 4 above)
Comment (rhugonnet) [reply geoutils/raster/raster.py:3545]: It will soon be with #492

Comment (rhugonnet) [reply geoutils/raster/raster.py:3545]: (and it is specified at the start of the function description)

File: geoutils/raster/raster.py (hunk 5 above)
Comment (rhugonnet) [reply geoutils/raster/raster.py:3552]: Done

File: geoutils/raster/raster.py (hunk 6 above)
Comment (rhugonnet) [reply geoutils/raster/raster.py:3581]: Yes, done!

File: tests/test_raster.py (hunk 7 above)
Comment (rhugonnet) [reply tests/test_raster.py:3010]: Good catch

File: tests/test_raster.py (hunk 8 above)
Comment (rhugonnet) [reply tests/test_raster.py:3014]: No as it didn't use ij2xy consistently yet, this was corrected in #509. I take a note and I'll see if it makes sense to change when merging conflicts.

File: geoutils/raster/raster.py (hunk 4 above)
Comment (adehecq) [reply geoutils/raster/raster.py]: In that case, should it not be "PointCloud" instead of "point cloud"?

Comment (rhugonnet) [reply geoutils/raster/raster.py]: All functions descriptions use lower-case to describe objects since 6 months ("array", "raster", "dataframe", etc) to match what is done in other packages (have lighter but still understand descriptipons).
And the typing points exactly to the class so the real object is unmistakable!
